fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+    *
+    * @param[in] U U value at the current integration point
+    * @param[in] normal normal vector (usually not a unit vector)
+    * @param[in] Tr element transformation
+    * @param[out] FUdotN normal flux from the given element at the current
+    * integration point
+    * @return double maximum characteristic velocity
+    */
+   virtual double ComputeFluxDotN(const Vector &U, const Vector &normal,
+                                  ElementTransformation &Tr, Vector &FUdotN)
+   {
+      double val = ComputeFlux(U, Tr, flux);
+      flux.Mult(normal, FUdotN);
+      return val;
+   }
+
+   /**
+    * @brief Construct a new Hyperbolic Form Integrator object
+    *
+    * @param[in] rsolver_ numerical flux
+    * @param[in] dim physical dimension
+    * @param[in] num_equations_ the number of equations
+    * @param[in] IntOrderOffset_ 2*p+IntOrderOffset order Gaussian quadrature
+    * will be used
+    */
+   HyperbolicFormIntegrator(RiemannSolver *rsolver_, const int dim,
+                            const int num_equations_,
+                            const int IntOrderOffset_ = 3)
+      : NonlinearFormIntegrator(),
+        num_equations(num_equations_),
+        IntOrderOffset(IntOrderOffset_),
+        rsolver(rsolver_)
+   {
+#ifndef MFEM_THREAD_SAFE
+      state.SetSize(num_equations);
+      flux.SetSize(num_equations, dim);
+      state1.SetSize(num_equations);
+      state2.SetSize(num_equations);
+      fluxN1.SetSize(num_equations);
+      fluxN2.SetSize(num_equations);
+      fluxN.SetSize(num_equations);
+      nor.SetSize(dim);
+#endif
+   }
+   /**
+    * @brief Construct an object with a fixed integration rule
+    *
+    * @param[in] rsolver_ numerical flux
+    * @param[in] dim physical dimension
+    * @param[in] num_equations_ the number of equations
+    * @param[in] ir integration rule to be used
+    */
+   HyperbolicFormIntegrator(RiemannSolver *rsolver_, const int dim,
+                            const int num_equations_,
+                            const IntegrationRule *ir)
+      : NonlinearFormIntegrator(ir),
+        num_equations(num_equations_),
+        IntOrderOffset(0),
+        rsolver(rsolver_)
+   {
+#ifndef MFEM_THREAD_SAFE
+      state.SetSize(num_equations);
+      flux.SetSize(num_equations, dim);
+      state1.SetSize(num_equations);
+      state2.SetSize(num_equations);
+      fluxN1.SetSize(num_equations);
+      fluxN2.SetSize(num_equations);
+      fluxN.SetSize(num_equations);
+      nor.SetSize(dim);
+#endif
+   }
+
+   /**
+    * @brief Get the element integration rule based on IntOrderOffset, @see
+    * AssembleElementVector. Used only when ir is not provided
+    *
+    * @param[in] el given finite element space
+    * @return const IntegrationRule& with order 2*p + IntOrderOffset
+    */
+   const IntegrationRule &GetRule(const FiniteElement &el)
+   {
+      int order;
+      order = 2 * el.GetOrder() + IntOrderOffset;
```

Comment:
```suggestion
      const int order = 2 * el.GetOrder() + IntOrderOffset;
```

Also, should the mesh Jacobian order be taken into account?